## Importing Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

## Import Dataframe

In [2]:
# define a path folder
path = r'C:\Users\jeanclaude\Instacart Basket Analysis'

In [3]:
# Import ords_prods_merged
df_ords_prods_merged =  pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data','orders_products_merged_updated.pkl'))

In [4]:
# Create subset of first 1M rows
df = df_ords_prods_merged[:1000000]

In [5]:
# Check the shape 
df.shape

(1000000, 20)

In [6]:
# Check the first 10 records 
df.head(10)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   
5   3367565        1    prior             6                   2   
6    550135        1    prior             7                   1   
7   3108588        1    prior             8                   1   
8   2295261        1    prior             9                   1   
9   2550362        1    prior            10                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   
5                  7                    19.0         196                  1   
6                  9                    20.0         196                  1   
7                 14                    14.0         196                  2   
8                 16                     0.0         196                  4   
9                  8                    30.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   
5          1         Soda        77              7     9.0   both   
6          1         Soda        77              7     9.0   both   
7          1         Soda        77              7     9.0   both   
8          1         Soda        77              7     9.0   both   
9          1         Soda        77              7     9.0   both   

   price_range_loc    price_range_loc     busiest_day     busiest_days  \
0              NaN  Mid-range product  Regularly busy   Regularly busy   
1              NaN  Mid-range product  Regularly busy  Least busy days   
2              NaN  Mid-range product  Regularly busy  Least busy days   
3              NaN  Mid-range product      Least busy  Least busy days   
4              NaN  Mid-range product      Least busy  Least busy days   
5              NaN  Mid-range product  Regularly busy   Regularly busy   
6              NaN  Mid-range product  Regularly busy     Busiest days   
7              NaN  Mid-range product  Regularly busy     Busiest days   
8              NaN  Mid-range product  Regularly busy     Busiest days   
9              NaN  Mid-range product      Least busy  Least busy days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2           Most orders  
3        Average orders  
4           Most orders  
5        Average orders  
6           Most orders  
7           Most orders  
8           Most orders  
9        Average orders

## Grouping data with pandas

In [7]:
# Group products by name
df.groupby('product_name')

## Aggregating Data with agg()

### Performing a Single Aggregation

In [8]:
# groupby() function is being assigned to the df dataframe creating the pandas object for “department_id.”
# Onto this resulting object, the agg() function is applied.
# The agg() function will return the mean of the given column, in this case, “order_number.” 

df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [9]:
# Aggregation conducted in a different way
df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

### Performing Multiple Aggregations

In [10]:
# All it comes down to is adding more arguments to your code.

df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

## Aggregating Data with transform()

In [11]:
df_ords_prods_merged['max_order'] = df_ords_prods_merged.groupby(['user_id'])['order_number'].transform(np.max)

In [12]:
df_ords_prods_merged.head(100)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
..       ...      ...      ...           ...                 ...   
95   3226575      360    prior             1                   5   
96   1469869      377    prior             3                   5   
97   1927023      387    prior             2                   4   
98    858092      420    prior             4                   1   
99    119314      420    prior             5                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
..                ...                     ...         ...                ...   
95                 12                     NaN         196                  1   
96                 17                     3.0         196                  9   
97                 10                    22.0         196                  3   
98                 19                    30.0         196                  2   
99                 19                    16.0         196                  2   

    reordered  ... aisle_id  department_id  prices  _merge  price_range_loc  \
0           0  ...       77              7     9.0    both              NaN   
1           1  ...       77              7     9.0    both              NaN   
2           1  ...       77              7     9.0    both              NaN   
3           1  ...       77              7     9.0    both              NaN   
4           1  ...       77              7     9.0    both              NaN   
..        ...  ...      ...            ...     ...     ...              ...   
95          0  ...       77              7     9.0    both              NaN   
96          0  ...       77              7     9.0    both              NaN   
97          0  ...       77              7     9.0    both              NaN   
98          0  ...       77              7     9.0    both              NaN   
99          1  ...       77              7     9.0    both              NaN   

      price_range_loc     busiest_day     busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy   Regularly busy        Average orders   
1   Mid-range product  Regularly busy  Least busy days        Average orders   
2   Mid-range product  Regularly busy  Least busy days           Most orders   
3   Mid-range product      Least busy  Least busy days        Average orders   
4   Mid-range product      Least busy  Least busy days           Most orders   
..                ...             ...              ...                   ...   
95  Mid-range product  Regularly busy   Regularly busy           Most orders   
96  Mid-range product  Regularly busy   Regularly busy        Average orders   
97  Mid-range product      Least busy  Least busy days           Most orders   
98  Mid-range product  Regularly busy     Busiest days        Average orders   
99  Mid-range product  Regularly busy  Least busy days        Average orders   

   max_order  
0         10  
1         10  
2         10  
3         10  
4         10  
..       ...  
95         3  
96         3  
97         8  
98        22  
99        22  

[100 rows x 21 columns]

In [13]:
# flags that assigns a “loyalty” label to a user ID based on its corresponding max order value

df_ords_prods_merged.loc[df_ords_prods_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
df_ords_prods_merged.loc[(df_ords_prods_merged['max_order'] <= 40) & (df_ords_prods_merged['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
df_ords_prods_merged.loc[df_ords_prods_merged['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [14]:
# Check the frequency of new "Loyalty_flag"
df_ords_prods_merged['loyalty_flag'].value_counts(dropna = False)

Regular customer    15891077
Loyal customer      10293737
New customer         6249398
Name: loyalty_flag, dtype: int64

In [15]:
# Check the head
df_ords_prods_merged[['user_id', 'loyalty_flag', 'order_number', 'max_order']].head(60)

user_id      loyalty_flag  order_number  max_order
0         1      New customer             1         10
1         1      New customer             2         10
2         1      New customer             3         10
3         1      New customer             4         10
4         1      New customer             5         10
5         1      New customer             6         10
6         1      New customer             7         10
7         1      New customer             8         10
8         1      New customer             9         10
9         1      New customer            10         10
10       15  Regular customer            15         22
11       15  Regular customer            17         22
12       15  Regular customer            18         22
13       15  Regular customer            21         22
14       15  Regular customer            22         22
15       19      New customer             2          9
16       19      New customer             5          9
17       19      New customer             7          9
18       21  Regular customer            10         33
19       31  Regular customer            10         20
20       31  Regular customer            17         20
21       43  Regular customer             6         11
22       43  Regular customer             9         11
23       52  Regular customer             1         27
24       52  Regular customer             2         27
25       52  Regular customer             3         27
26       52  Regular customer             5         27
27       52  Regular customer             7         27
28       52  Regular customer             9         27
29       52  Regular customer            10         27
30       52  Regular customer            11         27
31       52  Regular customer            12         27
32       52  Regular customer            15         27
33       52  Regular customer            17         27
34       52  Regular customer            18         27
35       52  Regular customer            19         27
36       52  Regular customer            21         27
37       67  Regular customer             1         24
38       67  Regular customer             2         24
39       67  Regular customer             3         24
40       67  Regular customer             4         24
41       67  Regular customer             5         24
42       67  Regular customer             6         24
43       67  Regular customer             8         24
44       67  Regular customer             9         24
45       67  Regular customer            10         24
46       67  Regular customer            11         24
47       67  Regular customer            12         24
48       67  Regular customer            13         24
49       67  Regular customer            14         24
50       67  Regular customer            16         24
51       67  Regular customer            18         24
52       67  Regular customer            19         24
53       67  Regular customer            21         24
54       67  Regular customer            22         24
55       67  Regular customer            23         24
56       81      New customer             3          7
57       81      New customer             5          7
58       82  Regular customer             1         19
59       82  Regular customer             3         19

## Step 2

In [16]:
# In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” 
# for a subset of your dataframe. Now, repeat this process for the entire dataframe.

# Group by "product_name" 
df_ords_prods_merged.groupby('product_name')

In [17]:
df_ords_prods_merged.shape

(32434212, 22)

In [18]:
# Split the data into groups by "department_id" and then calculate the mean, min and max for each "order_number" column
df_ords_prods_merged.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.457687   1  99
2                17.277920   1  99
3                17.179756   1  99
4                17.811403   1  99
5                15.213779   1  99
6                16.439806   1  99
7                17.225773   1  99
8                15.340520   1  99
9                15.895474   1  99
10               20.197148   1  99
11               16.170828   1  99
12               15.887622   1  99
13               16.583304   1  99
14               16.757377   1  99
15               16.165037   1  99
16               17.663250   1  99
17               15.694469   1  99
18               19.310397   1  99
19               17.177343   1  99
20               16.473447   1  99
21               22.902379   1  99

## Step 3

department_id 
4 went from a mean of 18.82 to 17.81
7 went from a mean of 17.47 to 17.22
13 went from a mean of 17.99 to 16.58
14 went from a mean of 19.24 to 16.75

besides the averages changing drastically on some departments like department_id 14(breakfast) and some not that much like department_id 7 (beverages)

The subset data only listed 8 department id's whereas the entire dataframe listed ALL 21 department id's. The subset therefore I would say would not be great to work with as it has tpoo many missing values.

## Step 4

In [19]:
# Group data by "user_id" column, generate mean product price for each user, create "mean_price" for aggregation results
df_ords_prods_merged['mean_price'] = df_ords_prods_merged.groupby(['user_id'])['prices'].transform(np.mean)

In [20]:
# Check grouping
df_ords_prods_merged.head(100)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
..       ...      ...      ...           ...                 ...   
95   3226575      360    prior             1                   5   
96   1469869      377    prior             3                   5   
97   1927023      387    prior             2                   4   
98    858092      420    prior             4                   1   
99    119314      420    prior             5                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
..                ...                     ...         ...                ...   
95                 12                     NaN         196                  1   
96                 17                     3.0         196                  9   
97                 10                    22.0         196                  3   
98                 19                    30.0         196                  2   
99                 19                    16.0         196                  2   

    reordered  ... prices  _merge   price_range_loc    price_range_loc  \
0           0  ...    9.0    both               NaN  Mid-range product   
1           1  ...    9.0    both               NaN  Mid-range product   
2           1  ...    9.0    both               NaN  Mid-range product   
3           1  ...    9.0    both               NaN  Mid-range product   
4           1  ...    9.0    both               NaN  Mid-range product   
..        ...  ...    ...     ...               ...                ...   
95          0  ...    9.0    both               NaN  Mid-range product   
96          0  ...    9.0    both               NaN  Mid-range product   
97          0  ...    9.0    both               NaN  Mid-range product   
98          0  ...    9.0    both               NaN  Mid-range product   
99          1  ...    9.0    both               NaN  Mid-range product   

       busiest_day     busiest_days busiest_period_of_day max_order  \
0   Regularly busy   Regularly busy        Average orders        10   
1   Regularly busy  Least busy days        Average orders        10   
2   Regularly busy  Least busy days           Most orders        10   
3       Least busy  Least busy days        Average orders        10   
4       Least busy  Least busy days           Most orders        10   
..             ...              ...                   ...       ...   
95  Regularly busy   Regularly busy           Most orders         3   
96  Regularly busy   Regularly busy        Average orders         3   
97      Least busy  Least busy days           Most orders         8   
98  Regularly busy     Busiest days        Average orders        22   
99  Regularly busy  Least busy days        Average orders        22   

        loyalty_flag mean_price  
0       New customer   6.367797  
1       New customer   6.367797  
2       New customer   6.367797  
3       New customer   6.367797  
4       New customer   6.367797  
..               ...        ...  
95      New customer  10.006250  
96      New customer   8.496552  
97      New customer   7.396610  
98  Regular customer   7.657813  
99  Regular customer   7.657813  

[100 rows x 23 columns]

In [21]:
# Create flags to assign "spending" labels to user ID based on mean product price

df_ords_prods_merged.loc[df_ords_prods_merged['mean_price'] < 10, 'spending_flag'] = 'Low Spending'
df_ords_prods_merged.loc[df_ords_prods_merged['mean_price'] >= 10, 'spending_flag'] = 'High Spending'

In [22]:
# Check frequencies of spending flags for accuracy
df_ords_prods_merged['spending_flag'].value_counts(dropna = False)

Low Spending     31798751
High Spending      635461
Name: spending_flag, dtype: int64

In [1]:
# Check sample data
df_ords_prods_merged[['user_id', 'loyalty_flag', 'order_number', 'max_order']].sample(60)

NameError: name 'df_ords_prods_merged' is not defined

## Step 5

In [24]:
#Check the basic statistics of product prices fot each of the loyalty categories
df_ords_prods_merged.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.388747  1.0  99999.0
New customer      13.294370  1.0  99999.0
Regular customer  12.496203  1.0  99999.0

From the output we can see that Loyal customers have a lower average spend per product ordered than New or Regular customers. There could be many possible reasons for this as the min and max range for products are between 1 and 99 and the products bought by the different categories could also differ

## Step 6

In [25]:
# Creating “average_price” column using transform()

df_ords_prods_merged['average_price'] = df_ords_prods_merged.groupby(['user_id'])['prices'].transform(np.mean)

In [26]:
df_ords_prods_merged.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered  ...  price_range_loc    price_range_loc     busiest_day  \
0          0  ...              NaN  Mid-range product  Regularly busy   
1          1  ...              NaN  Mid-range product  Regularly busy   
2          1  ...              NaN  Mid-range product  Regularly busy   
3          1  ...              NaN  Mid-range product      Least busy   
4          1  ...              NaN  Mid-range product      Least busy   

      busiest_days busiest_period_of_day max_order  loyalty_flag mean_price  \
0   Regularly busy        Average orders        10  New customer   6.367797   
1  Least busy days        Average orders        10  New customer   6.367797   
2  Least busy days           Most orders        10  New customer   6.367797   
3  Least busy days        Average orders        10  New customer   6.367797   
4  Least busy days           Most orders        10  New customer   6.367797   

  spending_flag average_price  
0  Low Spending      6.367797  
1  Low Spending      6.367797  
2  Low Spending      6.367797  
3  Low Spending      6.367797  
4  Low Spending      6.367797  

[5 rows x 25 columns]

In [27]:
# Creating 'spending_flag' 

df_ords_prods_merged.loc[df_ords_prods_merged['average_price'] < 10, 'spending_flag'] = 'Low spender'
df_ords_prods_merged.loc[df_ords_prods_merged['average_price'] >= 10, 'spending_flag'] = 'High spender'

In [28]:
df_ords_prods_merged.sample(1000)

order_id  user_id eval_set  order_number  orders_day_of_week  \
14045227   2992869   172357    prior            80                   0   
26260730   2520171    21185    prior             3                   0   
7140384    2317363   104835    prior             4                   5   
12217477   1789617   130624    prior             4                   0   
9639287    2383582    20771    prior            37                   3   
...            ...      ...      ...           ...                 ...   
14823355    758710    51594    prior            74                   6   
12812437    416475    17224    prior            13                   6   
31786709    510597   139124    prior            29                   6   
2226733     290044    15740    prior             1                   3   
5972262    2482841    45746    prior             7                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
14045227                 11                     3.0       19048   
26260730                 14                    30.0        4429   
7140384                  18                    28.0       44910   
12217477                 10                     8.0        4605   
9639287                  12                     1.0       46667   
...                     ...                     ...         ...   
14823355                  9                     1.0       43768   
12812437                  7                    12.0       33846   
31786709                 15                     5.0       21498   
2226733                  12                     NaN       47144   
5972262                   9                    30.0       28985   

          add_to_cart_order  reordered  ...  price_range_loc  \
14045227                  1          1  ...              NaN   
26260730                  1          1  ...              NaN   
7140384                  13          0  ...              NaN   
12217477                  5          1  ...              NaN   
9639287                   3          1  ...              NaN   
...                     ...        ...  ...              ...   
14823355                  2          1  ...              NaN   
12812437                  3          1  ...              NaN   
31786709                 11          1  ...              NaN   
2226733                   6          0  ...              NaN   
5972262                   2          1  ...              NaN   

            price_range_loc     busiest_day     busiest_days  \
14045227  Low-range product     Busiest day     Busiest days   
26260730  Mid-range product     Busiest day     Busiest days   
7140384   Mid-range product  Regularly busy   Regularly busy   
12217477  Low-range product     Busiest day     Busiest days   
9639287   Mid-range product  Regularly busy  Least busy days   
...                     ...             ...              ...   
14823355  Mid-range product  Regularly busy   Regularly busy   
12812437  Mid-range product  Regularly busy   Regularly busy   
31786709  Mid-range product  Regularly busy   Regularly busy   
2226733   Mid-range product  Regularly busy  Least busy days   
5972262   Mid-range product     Busiest day     Busiest days   

         busiest_period_of_day max_order      loyalty_flag mean_price  \
14045227           Most orders        99    Loyal customer   7.294493   
26260730           Most orders         4      New customer   7.972973   
7140384         Average orders        27  Regular customer   8.248113   
12217477           Most orders         6      New customer   8.647059   
9639287            Most orders        46    Loyal customer   8.205861   
...                        ...       ...               ...        ...   
14823355           Most orders        99    Loyal customer   7.279155   
12812437        Average orders        53    Loyal customer   7.988316   
31786709           Most orders        29  Regular customer   8.172707   
2226733            Most orders        14  Regula

## Step 7

In [29]:
# Creating “median_prior_orders” column using transform()

df_ords_prods_merged['median_prior_orders'] = df_ords_prods_merged.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [30]:
df_ords_prods_merged.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered  ...    price_range_loc     busiest_day     busiest_days  \
0          0  ...  Mid-range product  Regularly busy   Regularly busy   
1          1  ...  Mid-range product  Regularly busy  Least busy days   
2          1  ...  Mid-range product  Regularly busy  Least busy days   
3          1  ...  Mid-range product      Least busy  Least busy days   
4          1  ...  Mid-range product      Least busy  Least busy days   

   busiest_period_of_day max_order  loyalty_flag mean_price spending_flag  \
0         Average orders        10  New customer   6.367797   Low spender   
1         Average orders        10  New customer   6.367797   Low spender   
2            Most orders        10  New customer   6.367797   Low spender   
3         Average orders        10  New customer   6.367797   Low spender   
4            Most orders        10  New customer   6.367797   Low spender   

  average_price median_prior_orders  
0      6.367797                20.5  
1      6.367797                20.5  
2      6.367797                20.5  
3      6.367797                20.5  
4      6.367797                20.5  

[5 rows x 26 columns]

In [31]:
# Creating a 'order_frequency_flag' flag 'Non-frequent customer'

df_ords_prods_merged.loc[df_ords_prods_merged['median_prior_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [32]:
# Creating a 'order_frequency_flag' flag 'Regular customer'
df_ords_prods_merged.loc[(df_ords_prods_merged['median_prior_orders'] > 10) & (df_ords_prods_merged['median_prior_orders'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [33]:
# Creating a 'order_frequency_flag' flag 'Frequent customer'

df_ords_prods_merged.loc[df_ords_prods_merged['median_prior_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [34]:
df_ords_prods_merged.sample(1000)

order_id  user_id eval_set  order_number  orders_day_of_week  \
5890309    3392181   190569    prior            13                   1   
15657656   3197494   134078    prior             1                   4   
8065989    2288794   150691    prior            14                   2   
26848135   3099617    40601    prior             3                   0   
23515459    618777   161408    prior             2                   6   
...            ...      ...      ...           ...                 ...   
27132341   1798830    56911    prior            14                   6   
19676999   3343384   157924    prior             9                   4   
31389819    845618   187362    prior            77                   3   
32228956    359336    31665    prior             1                   0   
29807596   2181296   122810    prior             4                   2   

          order_hour_of_day  days_since_prior_order  product_id  \
5890309                  17                    11.0       26209   
15657656                 20                     NaN       20339   
8065989                  23                    10.0       28465   
26848135                 14                    28.0       48186   
23515459                 21                     7.0        1979   
...                     ...                     ...         ...   
27132341                 11                    26.0       17568   
19676999                 12                     4.0       35898   
31389819                 16                     4.0       17434   
32228956                 17                     NaN       44459   
29807596                 15                    16.0        8042   

          add_to_cart_order  reordered  ...     busiest_day     busiest_days  \
5890309                   2          1  ...  Regularly busy     Busiest days   
15657656                  1          0  ...      Least busy  Least busy days   
8065989                   3          1  ...  Regularly busy   Regularly busy   
26848135                  9          1  ...     Busiest day     Busiest days   
23515459                  9          0  ...  Regularly busy   Regularly busy   
...                     ...        ...  ...             ...              ...   
27132341                  6          1  ...  Regularly busy   Regularly busy   
19676999                  9          0  ...      Least busy  Least busy days   
31389819                  7          1  ...  Regularly busy  Least busy days   
32228956                  3          0  ...     Busiest day     Busiest days   
29807596                 27          0  ...  Regularly busy   Regularly busy   

          busiest_period_of_day  max_order      loyalty_flag mean_price  \
5890309          Average orders         13  Regular customer   8.444765   
15657656         Average orders         20  Regular customer   6.083959   
8065989           Fewest orders         43    Loyal customer   8.239050   
26848135            Most orders         11  Regular customer   8.084756   
23515459         Average orders          7      New customer   6.312308   
...                         ...        ...               ...        ...   
27132341            Most orders         16  Regular customer   7.688372   
19676999            Most orders         46    Loyal customer   7.304805   
31389819            Most orders         99    Loyal customer   8.076582   
32228956         Average orders          4      New customer   9.192308   
29807596            Most orders          9      New customer   8.189407   

         spending_flag average_price median_prior_orders  \
5890309    Low spender      8.444765                14.0   
15657656   Low spender      6.083959                17.0   
8065989    Low spender      8.239050                 6.0   
26848135   Low spender      8.084756                25.0   
23515459   Low spender      6.312308                 7.0   
...                ...           ...                 ...   
27132341   Low spender      7.6883

## Step 8

Code is commented and labeld accordingly

## Step 9 

In [37]:
# Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.
df_ords_prods_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_updated_2.pkl'))